File: 2025-9322-team04_finproject_python/Client_Python/player/controller/WaitingRoomController.py
# waiting_room_controller.py
import time
from Client_Python.manager.PlayerCorbaManager import PlayerCorbaManager
import threading
from Client_Python.player.main import HomepageMain, WaitingRoomMain
from Client_Python.player.main import GameProperMain
from Client_Python.player.model.WaitingRoomModel import WaitingRoomModel
from Client_Python.compilations.GameService_idl import _0_compilations
NoOpponentFoundException = _0_compilations.NoOpponentFoundException
GameNotFoundException = _0_compilations.GameNotFoundException

class WaitingRoomController:
    def __init__(self, username):
        self.model = WaitingRoomModel(username)
        orb_args = ["-ORBInitRef", "NameService=corbaloc::localhost:4321/NameService"]
        self.manager = PlayerCorbaManager.get_instance(orb_args)
        self.game_service = self.manager.get_game_service()
        self.current_round = 1
        self.game_started = False  # shared flag to stop all threads once started
        self.stop_threads = False

    #called by start_polling function from WaitingRoomMain
    def polling_loop(self):
        try:
            self.model.request_to_join_game(self.model.get_username())
            while True:
                status = self.model.get_game_status(self.model.get_username())
                if status.upper() == "STARTED":
                    self.game_started = True #sets to true
                    self.stop_threads = True

                    #open game proper
                    game = GameProperMain(
                        self.current_round,
                        self.model.get_username()
                    )
                    #print("Starting game Loop...")
                    game.run_round()
                    #print("Game Loop ended.") #DEBUG
                    break
                time.sleep(1)

        except NoOpponentFoundException:
            print("No opponent found. Try again.")
            self.stop_threads = True
            HomepageMain(self.model.get_username())
        except GameNotFoundException:
            print("Game not found.")
            self.stop_threads = True
            HomepageMain(self.model.get_username())
        except Exception as e:
            print(f"Error: {str(e)}")
            self.stop_threads = True
            HomepageMain(self.model.get_username())

    #called by start_countdown function from WaitingRoomMain
    def countdown(self):
        time_left = self.model.get_count_down_time()
        while time_left > 0:
            #print("DEBUG: IF GAME HAS STARTED " + str(self.game_started))  #FOR DEBUG
            if self.game_started:
                return  # STOPS COUNTDOWN IF GAME HAS STARTED
            print(f"Countdown: {time_left} seconds")
            time.sleep(1)
            time_left -= 1
            print()#move to next line


        # Countdown finished
        print("⏰ Countdown finished. Waiting for game to start...")
        def check_game_status():
            print("CHECK GAME STATUS THREAD")#FOR DEBUG
            try:
                time.sleep(3)
                status = self.game_service.getGameProgressStatus(self.model.get_username())#CHECK
                print(f"Status DEBUG: {status}")
                if status.lower() != "started":
                    print("⚠ Not enough players to start the game.")
                    self.stop_threads = True
                    option = input("Would you like to try again? (y/n): ").strip().lower()
                    if option == 'y':
                        WaitingRoomMain(self.model.get_username())  # You need to define this
                    else:
                        HomepageMain(self.model.get_username())
            except Exception as ex:
                print(f"Connection error: {ex}")
                HomepageMain(self.model.get_username())

        # Start a new thread to check game status after countdown
        threading.Thread(target=check_game_status).start()

    #called by start_polling_player_count function from WaitingRoomMain
    def polling_player_count(self):
        while self.model.get_count_down_time() > 0 and not self.game_started and not self.stop_threads:
            try:
                count = self.game_service.getCurrentPlayerCount()
                print(f"\rPlayers Joined: {count}")

                time.sleep(1)
            except Exception:
                print(-1)

File: 2025-9322-team04_finproject_java/src/main/java/Client_Java/player/controller/WaitingRoomController.java
package Client_Java.player.controller;

import Client_Java.player.model.HomepageModel;
import Client_Java.player.model.WaitingRoomModel;
import Client_Java.player.view.HomepageView;
import Client_Java.player.view.WaitingRoomView;
import Client_Java.player.PlayerCorbaManager;
import compilations.GameNotFoundException;
import compilations.GameService;
import compilations.NoOpponentFoundException;
import compilations.WordMaskInfo;
import compilations.WordService;

import javax.swing.*;

public class WaitingRoomController {
    private final WaitingRoomModel model;
    private final WaitingRoomView view;
    private final Runnable returnAction;
    private GameService gameService;
    private final WordService wordService;
    private final int currentRound = 1;

    public WaitingRoomController(WaitingRoomModel model,
                                 WaitingRoomView view,
                                 Runnable returnAction, GameService gameService, WordService wordService) {
        this.model = model;
        this.view = view;
        this.returnAction = returnAction;

        // Initialize both services
        PlayerCorbaManager corbaManager = PlayerCorbaManager.getInstance();
        this.gameService = corbaManager.getGameService();
        this.wordService = corbaManager.getWordService();

        // Show the view (replaced showWaitingRoom() with showView())
        view.showView();
        setupActions();
        startCountdown();
        startPolling();
        startPollingPlayerCount();
    }
    private void setupActions() {
        view.getBackButton().addActionListener(e -> exitWaitingRoom());
        view.getMinimizeBtn().addActionListener(e -> view.setState(JFrame.ICONIFIED));
        view.getCloseBtn().addActionListener(e -> System.exit(0));
        view.getMaximizeBtn().addActionListener(e -> toggleMaximize());
    }

    private void exitWaitingRoom() {
        view.setVisible(false);
        if (returnAction != null) {
            returnAction.run();
        }
        view.dispose();
    }

    private void toggleMaximize() {
        int state = view.getExtendedState();
        boolean isCurrentlyMaximized = (state & JFrame.MAXIMIZED_BOTH) == JFrame.MAXIMIZED_BOTH;

        if (isCurrentlyMaximized) {
            view.setExtendedState(JFrame.NORMAL);
            view.setSize(1280, 720);
            view.setLocationRelativeTo(null);
        } else {
            view.setExtendedState(JFrame.MAXIMIZED_BOTH);
        }
    }

    private void startPolling() {
        new Thread(() -> {
            try {
                // Register player with CORBA game service
                gameService.requestToJoinGame(model.getUsername());

                while (true) {
                    try {
                        String status = gameService.getGameProgressStatus(model.getUsername());
                        SwingUtilities.invokeLater(() -> view.updateStatus("Status: " + status));

                        if ("STARTED".equalsIgnoreCase(status)) {
                            WordMaskInfo wordInfo = gameService.getCurrentWordMask(model.getUsername());
                            if (wordInfo == null || wordInfo.actualWord == null || wordInfo.actualWord.trim().isEmpty()) {
                                SwingUtilities.invokeLater(() -> {
                                    JOptionPane.showMessageDialog(null,
                                            "Server did not send a valid word. Please restart the game.");
                                    exitWaitingRoom();
                                });
                                break;
                            }

                            // Initialize game with CORBA services
                            SwingUtilities.invokeLater(() -> {
                                HomepageModel homepageModel = new HomepageModel();
                                homepageModel.setUsername(model.getUsername());
                                HomepageView homepageView = new HomepageView();

                                // Pass CORBA manager to HomepageController
                                HomepageController homepageController = new HomepageController(
                                        homepageModel,
                                        homepageView
                                );

                                homepageController.openGameProper(
                                        wordInfo.actualWord,
                                        currentRound,
                                        wordService,
                                        gameService,
                                        model.getUsername(),
                                        wordInfo.gameId
                                );
                            });

                            SwingUtilities.invokeLater(() -> view.setVisible(false));
                            break;
                        }
                        Thread.sleep(1000);
                    } catch (org.omg.CORBA.TIMEOUT e) {
                        SwingUtilities.invokeLater(() -> {
                            view.updateStatus("Connection timeout. Retrying...");
                        });
                        // Re-establish CORBA connection
                        String[] orbArgs = new String[0];
                        PlayerCorbaManager.getInstance().reconnect(orbArgs); //

                        this.gameService = PlayerCorbaManager.getInstance().getGameService();
                    }
                }
            } catch (NoOpponentFoundException e) {
                SwingUtilities.invokeLater(() -> view.updateStatus("No opponent found. Try again."));
            } catch (GameNotFoundException e) {
                SwingUtilities.invokeLater(() -> view.updateStatus("Game not found."));
            } catch (Exception e) {
                SwingUtilities.invokeLater(() -> {
                    view.updateStatus("Error: " + e.getMessage());
                    exitWaitingRoom();
                });
            }
        }).start();
    }
    private void showHomepage(String username) {
        HomepageModel homepageModel = new HomepageModel();
        homepageModel.setUsername(username);

        HomepageView homepageView = new HomepageView();

        HomepageController homepageController = new HomepageController(homepageModel, homepageView);

        homepageView.show();
    }
    private void startCountdown() {
        new Thread(() -> {
            while (model.getCountdownTime() > 0) {
                int timeLeft = model.getCountdownTime();
                SwingUtilities.invokeLater(() -> view.updateCountdownLabel(timeLeft));
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                model.decrementCountdownTime();
            }

            SwingUtilities.invokeLater(() ->
                    view.updateStatus("⏰ Countdown finished. Waiting for game to start...")
            );

            new Thread(() -> {
                try {
                    Thread.sleep(3000);
                    String status = gameService.getGameProgressStatus(model.getUsername());
                    if (!"STARTED".equalsIgnoreCase(status)) {
                        SwingUtilities.invokeLater(() -> {
                            int option = JOptionPane.showConfirmDialog(
                                    null,
                                    "Not enough players to start the game. Would you like to try again?",
                                    "Game Not Started",
                                    JOptionPane.OK_CANCEL_OPTION,
                                    JOptionPane.WARNING_MESSAGE
                            );

                            if (option == JOptionPane.OK_OPTION) {
                                exitWaitingRoom(); // This will trigger the returnAction
                            } else {
                                exitWaitingRoom(); // Same behavior for cancel
                            }
                        });
                    }
                } catch (Exception ex) {
                    SwingUtilities.invokeLater(() -> {
                        JOptionPane.showMessageDialog(null,
                                "Connection error: " + ex.getMessage(),
                                "Error",
                                JOptionPane.ERROR_MESSAGE);
                        exitWaitingRoom();
                    });
                }
            }).start();
        }).start();
    }

    private void startPollingPlayerCount() {
        new Thread(() -> {
            while (model.getCountdownTime() > 0) {
                try {
                    short count = gameService.getCurrentPlayerCount();
                    SwingUtilities.invokeLater(() -> view.updatePlayerCount(count));
                    Thread.sleep(1000);
                } catch (Exception e) {
                    SwingUtilities.invokeLater(() -> view.updatePlayerCount(-1));
                }
            }
        }).start();
    }
}
